feat: Add a MongoDB template

sis0k0 opened the pull request:
This pull request introduces a new MongoDB template located in the `/mongodb` directory. The template provides three variations of environments:

- `/mongodb/blank` — A basic Node.js script with a MongoDB connection.
- `/mongodb/express` — An Express REST API.
- `/mongodb/flask` — A Flask REST API.

Each variation includes the `mongodb` service which sets up a local instance of MongoDB.

**/mongodb/[blank|express|flask]/.idx/dev.nix**
```
services.mongodb = {
  enable = true;
};
```

Each variation executes a script to start `mongodb` as a daemon when a new workspace is created.

**/mongodb/[blank|express|flask]/.idx/dev.nix**
```
onStart = {
  start-database = "mongod --port 27017 --fork --logpath ./.idx/database.log --dbpath ./.idx/.data";
};
```

Finally, each variation installs the MongoDB VSCode extension, enabling users to connect to the local database instance and interact with the data.

**/mongodb/[blank|express|flask]/.idx/dev.nix**
```
extensions = [
  "mongodb.mongodb-vscode"
];
```

<a href="https://idx.google.com/new?template=https%3A%2F%2Fgithub.com%2Fsis0k0%2Ftemplates%2Ftree%2Fmain%2Fmongodb">
  <img
    height="32"
    alt="Open in IDX"
    src="https://cdn.idx.dev/btn/open_purple_32.svg">
</a>

rodydavis commented:
When the Node.js first boots is there anything to do about the error state of not connecting to the local instance?

![Screenshot 2024-10-29 at 3 54 57 PM](https://github.com/user-attachments/assets/122ad548-464c-4f16-9f6a-53f3b2c8a7d2)

Clicking either of the options did connect:
![Screenshot 2024-10-29 at 3 56 47 PM](https://github.com/user-attachments/assets/49f44932-e361-4f94-961c-3727c3cd2460)

rodydavis commented:
Also is there a web preview you would want to show? Or only running via the terminal?
![Screenshot 2024-10-29 at 4 01 16 PM](https://github.com/user-attachments/assets/8159a5ef-902d-4618-b711-a1844c26f0aa)

rodydavis commented:
Probably expected, but running the example code twice gives an error:
![Screenshot 2024-10-29 at 4 02 53 PM](https://github.com/user-attachments/assets/7d9adccc-4fd7-43f4-ba29-bd674f47d5f8)

rodydavis commented on mongodb/blank/.idx/dev.nix:
```diff
@@ -0,0 +1,35 @@
+    };
+
+    previews = {
+      enable = true;
+      previews = {
+      };
```
This can be removed since it is not used in the template.

```suggestion
```

rodydavis commented on mongodb/blank/index.js:
```diff
@@ -0,0 +1,96 @@
+const { MongoClient } = require('mongodb');
```
Is there a ESM version of this? or require the default method?

sis0k0 replied:
I'll switch to importing the default export.

rodydavis commented:
For the express.js template did you want a web preview?

![Screenshot 2024-10-29 at 4 22 35 PM](https://github.com/user-attachments/assets/15883b5a-4ec6-4be4-b54d-442a625a2bab)

This would be the **dev.nix** for it:

```nix
{ pkgs, ... }: {
  channel = "stable-23.11";

  packages = [
    pkgs.nodejs_20
  ];

  services.mongodb.enable = true;

  idx = {
    extensions = [
      "mongodb.mongodb-vscode"
    ];

    workspace = {
      onCreate = {
        npm-install = "npm install";
        default.openFiles = [
          "server.js"
          "database.js"
          "README.md"
        ];
      };
      onStart = {
        start-database = "mongod --port 27017 --fork --logpath ./.idx/database.log --dbpath ./.idx/.data";
      };
    };

    previews = {
      enable = true;
      previews = {
        web = {
          command = [ "npm" "run" "start" ];
          manager = "web";
          env = {
            PORT = "$PORT";
          };
        };
      };
    };
  };
}
```

It just needs the **PORT** from the env:

```js
serverConfiguration: {
    port: process.env.PORT || 3000,
}
```

rodydavis commented on mongodb/express/.idx/dev.nix:
```diff
@@ -0,0 +1,36 @@
+    };
+
+    previews = {
+      enable = true;
+      previews = {
+      };
```
Can be removed if not being used:

```suggestion
```

rodydavis commented on mongodb/express/README.md:
````diff
@@ -0,0 +1,15 @@
+```
+
+1. Use the MongoDB extension from the left-hand sidebar to connect to the local database instance — `mongodb://localhost:27017`.
+1. Open the `server.js` file to view the implementation.
````
```suggestion
2. Open the `server.js` file to view the implementation.
```

rodydavis commented on mongodb/flask/.idx/dev.nix:
```diff
@@ -0,0 +1,41 @@
+    };
+
+    previews = {
+      enable = true;
+      previews = {
+      };
```
Can be removed if not being used

```suggestion
```

rodydavis commented on mongodb/flask/README.md:
````diff
@@ -0,0 +1,14 @@
+```
+
+1. Use the MongoDB extension from the left-hand sidebar to connect to the local database instance — `mongodb://localhost:27017`.
+1. Open the `app.py` file to view the implementation.
````
```suggestion
2. Open the `app.py` file to view the implementation.
```

rodydavis commented on mongodb/flask/README.md:
````diff
@@ -0,0 +1,14 @@
+The server should run automatically on port `5000` when starting a workspace. To run it manually, execute:
+
+```
+./devserver.sh
````
When you run the server on `http://127.0.0.1:5000` and open it up in the browser you get a 404.

![Screenshot 2024-10-29 at 4 31 22 PM](https://github.com/user-attachments/assets/58a07411-b6a0-42bd-85e5-685038ba186f)

This is because `http://127.0.0.1:5000/users` is a valid routed but does not have a fallback like the other examples.

sis0k0 replied:
I'll add a default '/' route.

rodydavis commented on mongodb/blank/README.md:
```diff
@@ -0,0 +1,9 @@
+## Getting Started
+
+1. Use the MongoDB extension from the left-hand sidebar to connect to the local database instance — `mongodb://localhost:27017`.
+1. Open `index.js` to view the implementation.
```
```suggestion
2. Open `index.js` to view the implementation.
```

rodydavis reviewed (changes requested):
Added some comments, but it is very close!

sis0k0 commented:
> When the Node.js first boots is there anything to do about the error state of not connecting to the local instance?
> 
> ![Screenshot 2024-10-29 at 3 54 57 PM](https://private-user-images.githubusercontent.com/31253215/381331327-122ad548-464c-4f16-9f6a-53f3b2c8a7d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNzEyMDYsIm5iZiI6MTczMDI3MDkwNiwicGF0aCI6Ii8zMTI1MzIxNS8zODEzMzEzMjctMTIyYWQ1NDgtNDY0Yy00ZjE2LTlmNmEtNTNmM2IyYzhhN2QyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDA2NDgyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNhZGM3MDc2NGQyNzgwNmEyMTA1NDI4YjViNmFjNjgzNzBiMzZjOTczNTA4OTBiMmI1ZDdiMGU5ZmVhN2I5YzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-3dfEHHQ3TLJ5KPO0KqsPrBZPzT6EkP2aBhLXSZy7nM)
> 
> Clicking either of the options did connect: ![Screenshot 2024-10-29 at 3 56 47 PM](https://private-user-images.githubusercontent.com/31253215/381331564-49f44932-e361-4f94-961c-3727c3cd2460.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNzEyMDYsIm5iZiI6MTczMDI3MDkwNiwicGF0aCI6Ii8zMTI1MzIxNS8zODEzMzE1NjQtNDlmNDQ5MzItZTM2MS00Zjk0LTk2MWMtMzcyN2MzY2QyNDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDA2NDgyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0YTRiYjAwYTYxZDJlOTdlOTJiYTAxN2UxYTY3ZWNjMjE1NDE1OTFkYTdjNDY0MGQ4MjVmOGQ1OTc0M2FjMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hu8_SsOvc9Q7HvTins1C2MPKi_iqGcSsGTylzxONqsg)

That's the welcome screen loaded by the MongoDB VSCode extension. Unfortunately, we can't programmatically setup a connection for the extension and the user needs to manually connect.

rodydavis commented:
Sounds good!

sis0k0 commented:
Hi @rodydavis! I have addressed the code suggestions. Regarding the other comments:

- Web previews: Actually, I don't think we need these if there is no GUI in the templates. What do you think?
- Duplicate email error when running the blank script twice: Yes, the error is expected, but now that I'm thinking about it, it might be confusing for the users. Would it be better to remove the schema validation from that template? What are your thoughts?

rodydavis commented:
> * Web previews: Actually, I don't think we need these if there is no GUI in the templates. What do you think?

Yes totally fine not to have it! If you had a client example for one of the templates (even in the future), then it would make sense.

> * Duplicate email error when running the blank script twice: Yes, the error is expected, but now that I'm thinking about it, it might be confusing for the users. Would it be better to remove the schema validation from that template? What are your thoughts?

It might just be more of the sample code that gets generated and maybe it could be a generated ID instead of a static one. It just threw me off guard and took a second to see if it was an error from something I did.

sis0k0 commented:
I like the idea of generating an ID, in this case — an email. I've added [`faker-js`](https://github.com/faker-js/faker) to create and insert fictional user data every time the script runs.

rodydavis commented:
Testing and reviewing the changes now!

rodydavis reviewed (approved):
LGTM
